File: src/FarmEngine.Schemas/Interface.cs
namespace FarmEngine.Schemas;

// Port of packages/engine-schemas/src/interface.ts.

/// <summary>TS <c>GamePanelSchema.entries[].kind</c> enum.</summary>
public static class GamePanelEntryKinds
{
    public const string Text = "text";
    public const string Money = "money";
    public const string Energy = "energy";
    public const string Day = "day";
    public const string Item = "item";
    public const string Flag = "flag";
    public const string Action = "action";

    public static readonly IReadOnlyList<string> All = [Text, Money, Energy, Day, Item, Flag, Action];
}

/// <summary>TS <c>GamePanelSchema.entries</c> item (inline object).</summary>
public sealed record GamePanelEntry
{
    public string Label { get; init; } = "";
    /// <summary>One of <see cref="GamePanelEntryKinds"/>.</summary>
    public string Kind { get; init; } = "";
    public string Value { get; init; } = "";
}

public sealed record GamePanel
{
    public string Id { get; init; } = "";
    public string Title { get; init; } = "";
    public string? VisibleFlag { get; init; }
    /// <summary>At most 40 entries.</summary>
    public List<GamePanelEntry> Entries { get; init; } = [];
}

File: crates/farm-sim/src/schema/interface.rs
//! Port of `Interface.cs` (packages/engine-schemas/src/interface.ts).

use serde::{Deserialize, Serialize};

/// TS `GamePanelSchema.entries[].kind` enum.
pub mod game_panel_entry_kinds {
    pub const TEXT: &str = "text";
    pub const MONEY: &str = "money";
    pub const ENERGY: &str = "energy";
    pub const DAY: &str = "day";
    pub const ITEM: &str = "item";
    pub const FLAG: &str = "flag";
    pub const ACTION: &str = "action";

    pub const ALL: &[&str] = &[TEXT, MONEY, ENERGY, DAY, ITEM, FLAG, ACTION];
}

/// TS `GamePanelSchema.entries` item (inline object).
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct GamePanelEntry {
    pub label: String,
    /// One of [`game_panel_entry_kinds`].
    pub kind: String,
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct GamePanel {
    pub id: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub visible_flag: Option<String>,
    /// At most 40 entries.
    pub entries: Vec<GamePanelEntry>,
}
